SALSA runs in publicly-available docker containers that contain all the necessary dependencies for code execution. This workflow assumes that you have already aligned your raw data with cellranger, cellranger-atac, or cellranger-arc to generate coordinate-sorted bam files. For additional information, please consult the 10X Genomics website: https://www.10xgenomics.com/

SALSA is run in two stages:
1. Generate phased and annotated single cell allele-specific counts
2. Analyze allele-specific counts 

STAGE 1: 
p4rkerw/salsa:count_1.0 has the following dependencies:
GATK 4.2.0.0
bwa
STAR
bcftools
hdf5 \
pysam
shapeit 4.2
WASP 0.3.4

**STEP 1: Download cellranger references** If you don't already have the reference files that were used to align your datasets then download them from the cellranger website. For example, if you are analyzing human single cell gene expression and ATAC data on GRCh38 you should have the following folders mounted to the rna_ref and atac_ref directories:
```
GRCh38-2020-A.premrna
refdata-cellranger-atac-GRCh38-1.2.0
```

**STEP 2: Download GATK resource bundle files** Download GATK resource bundle files from the GATK google cloud bucket: https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false

The following files are required for GATK HaplotypeCaller:
```
resources_broad_hg38_v0_Homo_sapiens_assembly38.dbsnp138.vcf.gz
resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz
resources_broad_hg38_v0_1000G_phase1.snps.high_confidence.hg38.vcf.gz
resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz
```

The following additional files are required if you are analyzing single cell ATAC or Multiome datasets:
```
resources_broad_hg38_v0_wgs_calling_regions.hg38
resources_broad_hg38_v0_hapmap_3.3.hg38.vcf.gz
```

**STEP 3: Pull stage 1 container and mount volumes** Pull the docker container from dockerhub and mount the required volumes in an interactive session. The $SCRATCH1 variable designates a temporary file directory. The following command mounts directories with counts generated by cellranger mounted to rna_counts and counts generated by cellranger-atac mounted to atac_counts. These directories contain folders with the cellranger (or cellranger-atac) output named by their library_id.

For analyzing single cell gene expression datasets:
```
SCRATCH1=/g/scratch
docker run --memory 64g
--workdir $HOME
-v $HOME:$HOME
-v path/to/cellranger_rna_counts:$HOME/rna_counts
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref
-v path/to/vcf_output:$HOME/vcfdir
-v path/to/SALSA:$HOME/SALSA
-v path/to/gatk_bundle_reference:$HOME/gatk_bundle
-v $SCRATCH1:$SCRATCH1
-e SCRATCH1="/g/scratch"
--rm -it p4rkerw/salsa:count_1.0
```

For analyzing single cell gene expression, single cell ATAC and single cell Multiome datasets mount additional volumes for the ATAC reference and counts:
```
SCRATCH1=/g/scratch
docker run --memory 64g
--workdir $HOME
-v $HOME:$HOME
-v path/to/cellranger_rna_counts:$HOME/rna_counts
-v path/to/cellranger_atac_counts:$HOME/atac_counts
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref
-v path/to/refdata-cellranger-atac-GRCh38-1.2.0:$HOME/atac_ref
-v path/to/vcf_output:$HOME/vcfdir
-v path/to/SALSA:$HOME/SALSA
-v path/to/gatk_bundle_reference:$HOME/gatk_bundle
-v $SCRATCH1:$SCRATCH1
-e SCRATCH1="/g/scratch"
--rm -it p4rkerw/salsa:count_1.0
```

**STEP 4a: Genotype a single chromosome for a single cell gene expression dataset** The library_id variable refers to the name of the output folder generated by cellranger count (or cellranger-atac count) located in the rna_counts (or atac_counts) directory.
```
library_id=sample_1
interval=chr10
modality=rna
bash diabeticKidney/allele_specific_analysis/step1_gatk_genotype.sh
--bam ${modality}_counts/$sample/outs/possorted*.bam \
--library_id $library_id
--outputdir vcfdir/${modality}_genotype
--outputvcf $library_id.$modality.$interval.vcf.gz
--interval $interval
--modality $modality
--threads 4
```

**STEP 4b: Genotype a single chromosome for a single cell ATAC dataset**
```
library_id=sample_1
interval=chr10
modality=atac
bash diabeticKidney/allele_specific_analysis/step1_gatk_genotype.sh
--bam ${modality}_counts/$sample/outs/possorted*.bam
--library_id $library_id
--outputdir vcfdir/${modality}_genotype
--outputvcf $library_id.$modality.$interval.vcf.gz
--interval $interval
--modality $modality
--threads 4
```

**(Optional) STEP 5:** If you genotyped a paired single cell gene expression and ATAC dataset (or a single cell Multiome) for the same patient you can merge these genotypes into a single vcf:
```
modalityone=atac
modalitytwo=rna
library_id=sample_1
bash diabeticKidney/allele_specific_analysis/step2_merge_geno.sh
--library_id $library_id
--vcfone vcfdir/${modalityone}_genotype/$library_id.$modalityone.$interval.vcf.gz
--vcftwo vcfdir/${modalitytwo}_genotype/$library_id.$modalitytwo.$interval.vcf.gz
--outputdir vcfdir/joint_genotype
--outputvcf $library_id.pass.joint.$interval.vcf.gz
--threads 4
```

**(Optional) STEP 6:** If you want to perform your analysis with phased genotypes you will need a phased reference. This is not stricly required, but it increases the performance of the WASP variant-realignment and ASEP analysis steps. Download the 1000G phased reference files for either SNV or SNV and indels from ftp.1000genomes.ebi.ac.uk . Navigate to the corresponding directory depending on your selected reference. If you are only analyzing RNA data then select the SNV reference. For ATAC or Multiome data select the SNV and INDEL reference:

a) SNV only: /vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV </br>
b) SNV and INDEL: /vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL

You will eventually need to download the vcf for every chromosome, but for the purposes of the tutorial just download the SNV reference for chromosome10:
```
ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
```
The 1000G do not have the same contig style as the cellranger references. Update the 1000G contig style using bcftools in the stage 1 container. For the tutorial we will only do chromosome 10.
```
for i in {1..22} X;do echo "${i} chr${i}";done > /tmp/rename_chrm.txt
inputvcf=ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
bcftools annotate phasing/$inputvcf --threads 4 --rename-chrs /tmp/rename_chrm.txt -Oz -o $SCRATCH1/$inputvcf
mv $SCRATCH1/$inputvcf phasing/
bcftools index --threads 4 phasing/$inputvcf
```
Mount the directory with the 1000G reference and phase the genotypes
```
SCRATCH1=/g/scratch
docker run
--workdir $HOME
-v $HOME:$HOME
-v g/diabneph/analysis/combined_adv/vcf_filtered:$HOME/vcfdir
-v g/diabneph/github_repository/diabeticKidney:$HOME/diabeticKidney
-v g/reference/phasing/biallelic_SNV/ucsc:$HOME/phasing
-v $SCRATCH1:$SCRATCH1
-e SCRATCH1="/g/scratch"
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
interval=chr10
bash diabeticKidney/allele_specific_analysis/step3_phase_vcf.sh
--library_id $library_id
--inputvcf vcfdir/joint_genotype/$library_id.pass.joint.$interval.vcf.gz
--outputdir vcfdir/phasing
--outputvcf $library_id.pass.joint.${interval}hcphase.vcf.gz
--interval $interval
--hcphase
--threads 4
--snvonly
```






