:hot_pepper: SALSA is a tool for generating and analyzing phased single cell allele-specific read counts. The SALSA workflow runs in publicly-available docker containers with all the necessary dependencies for code execution. This workflow assumes you have aligned your raw data with cellranger or cellranger-atac to generate coordinate-sorted bam files and analyzed it to obtain cell barcode annotations. For additional information, please consult the 10X Genomics website: https://www.10xgenomics.com/ . For this tutorial, we will use a dataset downloaded from the 10X Genomics website that has already been aligned and annotated. To complete the tutorial, you will also need to download the cellranger reference, GATK bundle resources, and 1000G phased references (see below for more information). 

There are two stages in the workflow:
1. Generate phased and annotated single cell allele-specific counts from a cellranger bam
2. Analyze allele-specific counts 

STAGE 1: Steps 1-7 use p4rkerw/salsa:count_1.0, which is built on the broadinstitue/gatk:4.2.0.0 docker image with additional dependencies pre-installed:
```
GATK 4.2.0.0
bwa 0.7.17
STAR 2.7.8a
bcftools 1.8 
pysam 0.15.3
shapeit 4.2
WASP 0.3.4
```

**STEP 0: Download cellranger references** If you don't already have a GRCh38 cellranger reference download it from the 10X Genomics website. 10X Genomics routinely updates their references with each new cellranger build. The GRCh38-2020-A cellranger reference in this tutorial was selected because it was used to align the pbmc dataset on the 10X Genomics website. Feel free to use a different reference and/or dataset from their [collection](https://support.10xgenomics.com/single-cell-gene-expression/datasets), but make sure that it's aligned to the GRCh38 build to match the GATK bundle resources. 
```
# cellranger reference for scRNA analysis GRCh38 / hg38
reference=/g/reference
wget -P $reference https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2020-A.tar.gz
tar -xvzf $reference/refdata-gex-GRCh38-2020-A.tar.gz
```

**STEP 0: Download GATK resource bundle files** The following files are required for GATK HaplotypeCaller using GRCh38 and can be found in the [GATK google cloud bucket](https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false):
```
resources_broad_hg38_v0_Homo_sapiens_assembly38.dbsnp138.vcf.gz
resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz
resources_broad_hg38_v0_1000G_phase1.snps.high_confidence.hg38.vcf.gz
resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz
```
The following additional files are required if you are analyzing single cell ATAC datasets (not needed for the tutorial):
```
resources_broad_hg38_v0_wgs_calling_regions.hg38
resources_broad_hg38_v0_hapmap_3.3.hg38.vcf.gz
```
**STEP 0: Pull the SALSA container** 
```
docker pull p4rkerw/salsa:count_1.0
```
**STEP 1: Genotype a single cell gene expression dataset** The library_id variable refers to the name of the output folder generated by cellranger count (or cellranger-atac count) located in the rna_counts (or atac_counts) directory. If you're looking for a smaller dataset to work with in this tutorial, there are lots of options on the [10X Genomics website](https://support.10xgenomics.com/single-cell-gene-expression/datasets). For the purposes of the tutorial, you may consider downloading the coordinate-sorted bam and index for a dataset composed of 1k PBMCs from a healthy donor:
```
# URL to the dataset: https://support.10xgenomics.com/single-cell-gene-expression/datasets/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/
# navigate to your project directory and download the bam files
project=$PWD/salsa
wget -P $project/cellranger_rna_counts https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam
wget -P $project/cellranger_rna_counts https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam.bai
```
**Clone the SALSA github repository**
```
git -C $project clone https://github.com/p4rkerw/SALSA
```
**Usage:**
```
Usage: step1_gatk_genotype.sh [-indomlt]
   -i  | --bam                STR   path/to/input.bam eg. [rna_counts/sample_1/outs/possorted*.bam]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -d  | --outputdir          STR   output directory name eg. [vcfdir/rna_genotype]
   -o  | --outputvcf          STR   name of output vcf eg. [sample_1.rna.vcf.gz]
   -m  | --modality           STR   sequencing modality for short variant discovery: [rna] [atac]
   -l  | --interval           STR   optional: genotype a single chromosome eg. [chr10]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container** : Mount the required volumes in an interactive session. The $SCRATCH1 variable designates a temporary file directory. The following command mounts directories with counts generated by cellranger mounted to rna_counts. Users may prefer to mount all the directories required for future steps (see below) when they run the container, but for the sake of simplicity only the required volumes are specified at each step.
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/cellranger_rna_counts:$HOME/rna_counts \
-v $reference/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $reference/gatk:$HOME/gatk_bundle \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Genotype an RNA sample**
```
bash SALSA/step1_gatk_genotype.sh \
--bam rna_counts/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam \
--library_id pbmc \
--outputdir vcfdir/rna_genotype \
--outputvcf pbmc.rna.chr10.vcf.gz \
--interval chr10 \
--modality rna \
--threads 4
```
**(Not required for tutorial) STEP 2: Merge genotypes from the same patient** If you genotyped a paired single cell gene expression and ATAC dataset from a split sample (or a single cell Multiome) you can merge these genotypes into a single vcf. If you're following the tutorial, you can skip this step.
**Usage**
```
Usage: step2_merge_geno.sh [-nabdit]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -a  | --vcfone             STR   path/to/first.vcf.gz eg. [vcfdir/atac_genotype/sample_1.atac.vcf.gz]. Prioritize annotations from first vcf if there is overlap.
   -b  | --vcftwo             STR   path/to/second.vcf.gz eg. [vcfdir/rna_genotype/sample_1.rna.vcf.gz]
   -d  | --outputdir          STR   output directory [vcfdir/joint_genotype]
   -o  | --outputvcf          STR   name of output vcf eg. [sample_1.pass.joint.vcf.gz]
   -i  | --includefiltered          optional: include filtered variants in output vcf. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container**
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Merge two genotypes**
```
bash SALSA/step2_merge_geno.sh \
--library_id sample_1 \
--vcfone vcfdir/$atac_genotype/sample_1.atac.chr10.vcf.gz \
--vcftwo vcfdir/${modalitytwo}_genotype/sample_1.rna.chr10.vcf.gz \
--outputdir vcfdir/joint_genotype \
--outputvcf sample_1.pass.joint.chr10.vcf.gz \
--threads 4
```
**(Recommended) STEP 3: Phase genotype** If you want to perform your analysis with phased genotypes you will need a phased reference. This is not stricly required, but it increases the performance of the WASP variant-realignment and ASEP analysis steps. Download the 1000G phased reference files for either SNV or SNV and indels from ftp.1000genomes.ebi.ac.uk . Navigate to the corresponding directory depending on your selected reference. If you are only analyzing RNA data then select the SNV reference. For ATAC data select the SNV and INDEL reference:

a) SNV only: /vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV </br>
b) SNV and INDEL: /vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL

You will eventually need to download the vcf for every chromosome, but for the purposes of the tutorial just download the SNV reference for chromosome10 to reference/phasing:
```
ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
```
**Usage**
```
Usage: step3_phase_vcf.sh [-nvdolpsitrh]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -v  | --inputvcf           STR   path/to/input.vcf.gz eg. [vcfdir/joint_genotype/sample_1.pass.joint.vcf.gz]
   -d  | --outputdir          STR   output directory name eg. [vcfdir/phasing]
   -o  | --outputvcf          STR   name of output vcf eg. [sample_1.pass.joint.phase.vcf.gz]
   -l  | --interval           STR   optional: phase a single chromosome eg. [chr22]
   -p  | --hcphase                  optional: recover haplotypecaller physical phasing variants that are not in shapeit reference. Default=[false]
   -s  | --snvonly            STR   use the biallelic_SNV reference for phasing
   -i  | --snvindel           STR   use the biallelic_SNV_and_INDEL reference for phasing
   -r  | --reproduce                optional: run shapeit with a single thread for reproducibility. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
The 1000G do not have the same contig style as the cellranger references. Update the 1000G contig style using bcftools in the stage 1 container. For the tutorial we will only do chromosome 10. Mount the directory with the 1000G reference and phase the genotypes. For the tutorial we do not set the --reproduce flag, which is required to set a seed for reproducible results with shapeit4.2. This step is recommended in the full workflow.

**Launch SALSA container**
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $reference/phasing/biallelic_SNV:$HOME/phasing \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Rename the reference contigs**
```
# rename the contigs in the 1000G reference from 10 to chr10
for i in {1..22} X;do echo "${i} chr${i}";done > /tmp/rename_chrm.txt
inputvcf=ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
bcftools annotate phasing/$inputvcf --threads 4 --rename-chrs /tmp/rename_chrm.txt -Oz -o $SCRATCH1/$inputvcf
mv $SCRATCH1/$inputvcf phasing/
bcftools index --threads 4 phasing/$inputvcf
```
**Phase an interval**
```
bash SALSA/step3_phase_vcf.sh \
--library_id pbmc \
--inputvcf vcfdir/rna_genotype/pbmc.rna.chr10.vcf.gz \
--outputdir vcfdir/phasing \
--outputvcf pbmc.pass.joint.chr10hcphase.vcf.gz \
--interval chr10 \
--hcphase \
--threads 4 \
--snvonly
```
**(Optional) STEP 4: Annotate vcf with GATK Funcotator** If you would like to annotate your vcf with gnomAD MAF you will first need to download the GATK Funcotator resource following the directions located here: https://gatk.broadinstitute.org/hc/en-us/articles/360035889931-Funcotator-Information-and-Tutorial . The gnomAD resources need to be enabled after download (see GATK instructions on their website). When the resources have been downloaded move the dataSources folder into to the reference directory (eg. [reference/funcotator_dataSources.v1.6.20190124])
**Usage**
```
Usage: step4_gatk_anno_vcf.sh [-nvdoamfth]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -v  | --inputvcf           STR   path/to/input.vcf.gz eg. [vcfdir/phasing/sample_1.pass.joint.hcphase.vcf.gz]
   -d  | --outputdir          STR   output directory name eg. [vcfdir/funcotation]
   -o  | --outputvcf          STR   name of output vcf eg. [sample_1.pass.joint.hcphase.funco.vcf.gz]
   -a  | --output_table       STR   name of output funcotation csv eg. [sample_1.pass.joint.hchpase.formatted.csv]
   -m  | --modality           STR   sequencing modality for short variant discovery: [rna] [atac]
   -f  | --funcotation        STR   path/to/funcotation directory eg. [reference/funcotator_dataSources.v1.6.20190124g]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container**
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $reference/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $reference:$HOME/reference \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Annotate a vcf**
```
bash SALSA/step4_gatk_anno_vcf.sh \
--library_id pbmc \
--inputvcf vcfdir/phasing/pbmc.pass.joint.chr10hcphase.vcf.gz \
--outputdir vcfdir/funcotation \
--outputvcf pbmc.pass.joint.chr10hcphase.funco.vcf.gz \
--output_table pbmc.pass.joint.chr10hcphase.formatted.csv \
--modality rna \
--threads 4 \
--funcotation reference/funcotator_dataSources.v1.6.20190124g
```
**(Recommended) STEP 5:** Use barcode celltype annotations to filter the coordinate-sorted cellranger bam using the CB tag. This step is required if you want to analyze cell-specific or single cell allelic effects. The barcode annotation file has three columns where the first column is the barcode, the second column is the library_id, and the third column is the celltype annotation. If you have analyzed your dataset using Seurat you could generate a barcode annotation csv in R as follows (the Seurat package is not included in the SALSA containers). Mount the barcodes directory and filter the cellranger bam using the 10X Genomics subset-bam utility. For the purposes of the tutorial, we will only filter chromosome 10. The --validate flag runs GATK ValidateSamFile on the output to ensure the bam file meets specifications. To filter an ATAC coordinate-sorted bam change the modality and library_id variables as needed.
**Usage**
```
Usage: step5_filterbam.sh [-nidolmbeth]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -i  | --inputbam           STR   path/to/input.bam eg. [rna_counts/sample_1/outs/possorted_genome_bam.bam]
   -d  | --outputdir          STR   output directory eg. [project/wasp_rna]
   -o  | --outputbam          STR   filtered output bam eg. [sample_1.bcfilter.bam]
   -l  | --interval           STR   optional: filter a single chromosome eg. [chr22]
   -m  | --modality           STR   sequencing modality for short variant discovery: [rna] [atac]
   -b  | --barcodes           STR   path/to/barcodes.csv with headers and three columns. First column is named "barcodes"
                                    second column is group "orig.ident" eg. [barcodes/rna_barcodes.csv]
   -e  | --validate                 validate the barcode-filtered bam file. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container**
```
SCRATCH1=/g/scratch
project=/g/salsa
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/cellranger_rna_counts:$HOME/rna_counts \
-v $project/SALSA:$HOME/SALSA \
-v $project/barcodes:$HOME/barcodes \
-v $project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Download clustering analysis for tutorial dataset and create a barcode csv**
```
# download the file
wget -P project https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_analysis.tar.gz
tar -C $project/cellranger_rna_counts -xvzf $project/SC3_v3_NextGem_DI_PBMC_CSP_1K_analysis.tar.gz

# use the cluster number as celltype and assign pbmc as the library_id in final csv
cluster=$project/cellranger_rna_counts/analysis/clustering/graphclust/clusters.csv
(echo "barcode,orig.ident,celltype"; awk 'BEGIN{FS=OFS=","} {if (NR!=1) print $1,"pbmc",$2}' $cluster) |sed 's/-1//g'> barcodes/rna_barcodes.csv
```
**Filter a cellranger bam with barcode csv**
```
bash SALSA/step5_filterbam.sh \
--library_id pbmc \
--validate \
--inputbam rna_counts/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam \
--modality rna \
--interval chr10 \
--barcodes barcodes/rna_barcodes.csv \
--threads 4 \
--outputdir project/wasp_rna \
--outputbam pbmc.bcfilter.chr10.bam 
```
**STEP 6: Perform variant-aware realignment with WASP** This step takes a genotyped vcf and performs variant-aware realignment on a coordinate-sorted and indexed bam file with WASP. WASP is a tool to perform unbiased allele-specific read mapping and you can read more about it here: https://github.com/bmvdgeijn/WASP . For the purposes of the tutorial, we will only analyze chromosome 10. For RNA analysis, this step requires a STAR index of the cellranger reference. A STAR index can be built ahead of time using the following command:
```
STAR
--runMode genomeGenerate \
--runThreadN $threads \
--genomeDir rna_ref/star \
--genomeFastaFiles rna_ref/fasta/genome.fa \
--sjdbGTFfile rna_ref/genes/genes.gtf
```
Alternatively, these references will be built at runtime and placed in the $SCRATCH directory.

**Usage**
```
Usage: step6_wasp.sh [-vbdoginlmpt]
   -v  | --inputvcf          STR   vcfdir/funcotation/sample_1.pass.joint.hcphase.funco.vcf.gz
   -b  | --inputbam          STR   path/to/input.bam eg. [project/wasp_rna/sample_1.bcfilter.bam]
   -d  | --outputdir         STR   name of output directory eg. [project/wasp_rna]
   -o  | --outputbam         STR   name of output wasp bam eg. [sample_1.phase.wasp.bam]
   -g  | --genotype          STR   genotype: [rna] [atac] [joint]
   -i  | --stargenome        STR   path/to/star genomeDir made with genomeGenerate eg. [rna_ref/star]. If absent create new STAR index in $SCRATCH1/rna_ref
   -n  | --library_id        STR   library_id: eg. [sample_1]
   -m  | --modality          STR   modality: [rna] [atac]
   -l  | --interval          STR   specified interval eg. [chr10]
   -p  | --isphased                input vcf is phased. Default=[FALSE]
   -t  | --threads           INT   number of threads. Default=[1]
   -h  | --help                    show usage
```
**Launch SALSA container**
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $reference/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v $reference:$HOME/reference \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Run WASP on the barcode-filtered bam**
```
bash SALSA/step6_wasp.sh \
--inputvcf vcfdir/funcotation/pbmc.pass.joint.chr10hcphase.funco.vcf.gz \
--inputbam project/wasp_rna/pbmc.bcfilter.chr10.bam \
--outputdir project/wasp_rna/rna_genotype \
--outputbam pbmc.hcphase.chr10wasp.bam \
--genotype joint \
--stargenome rna_ref/star \
--library_id pbmc \
--modality rna \
--threads 4 \
--isphased \
--interval chr10
```
**STEP 7: Obtain allele-specific read counts with GATK** This step will filter a phased and genotyped vcf for heterozygous SNV to perform allele-specific counting in a coordinate-sorted and indexed bam file after WASP realignment. There are multiple options for count table outputs. The --pseudobulk option will group all barcodes together to perform allele-specific counting. This is analogous to bulk RNA-seq. The --celltype_pseudobulk option will use the barcode annotations to split the bam into cell-type-specific bam files before performing allele-specific counting. The --single_cell_counts option will split the bam file into individual single cell bam files and perform allele-specific counting. If your input vcf is phased you can select the --isphased option to add a phased genotype to the count tables.

**Usage**
```
Usage: step7_gatk_alleleCount.sh [-viognmlCcspt]
   -v  | --inputvcf           STR   path/to/input.vcf.gz eg. [vcfdir/funcotation/sample_1.pass.joint.hcphase.funco.vcf.gz]
   -i  | --inputbam           STR   path/to/wasp.bam eg. [project/wasp_rna/sample_1.phase.wasp.bam]
   -o  | --outputdir          STR   path/to/output directory eg. [project/wasp_rna/counts]
   -b  | --barcodes           STR   path/to/barcodes.csv eg. [barcodes/rna_barcodes.csv]
   -g  | --genotype           STR   genotype: [rna] [atac] [joint]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -m  | --modality           STR   sequencing modality for short variant discovery: [rna] [atac]
   -l  | --interval           STR   optional: count a specified interval eg. [chr10]
   -C  | --pseudobulk_counts        allele-specific counts with all cells grouped together
   -c  | --celltype_counts          allele-specific counts after grouping cells by barcode annotation
   -s  | --single_cell_counts       single cell allele-specific counts for provided barcodes
   -p  | --isphased                 optional: input vcf is phased. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container**
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v path/to/refdata-cellranger-atac-GRCh38-1.2.0:$HOME/atac_ref \
-v path/to/vcf_output:$HOME/vcfdir \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/barcodes:$HOME/barcodes \
-v path/to/project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Get phased allele-specific counts**
```
library_id=sample_1
modality=rna
interval=chr10
bash diabeticKidney/allele_specific_analysis/step7_gatk_alleleCount.sh \
--inputvcf vcfdir/funcotation/$library_id.pass.joint.${interval}hcphase.funco.vcf.gz \
--inputbam project/wasp_${modality}/$library_id.phase.${interval}wasp.bam \
--outputdir project/wasp_${modality}/counts \
--barcodes barcodes/${modality}_barcodes.csv \
--genotype joint_genotype \
--library_id $library_id \
--modality $modality \
--pseudobulk_counts \
--single_cell_counts \
--celltype_counts \
--threads 4 \
--interval $interval \
--isphased
```








