:hot_pepper: SALSA runs in publicly-available docker containers with all the necessary dependencies for code execution. This workflow assumes that you have already aligned your raw data with cellranger or cellranger-atac to generate coordinate-sorted bam files. For additional information, please consult the 10X Genomics website: https://www.10xgenomics.com/ . To analyze individual cell types or single cells you will also need cell type barcode annotations.

There are two stages in the workflow:
1. Generate phased and annotated single cell allele-specific counts from a cellranger bam
2. Analyze allele-specific counts 

STAGE 1: 
p4rkerw/salsa:count_1.0 is built on the broadinstitue/gatk:4.2.0.0 docker image with additional dependencies pre-installed:
```
GATK 4.2.0.0
bwa 0.7.17
STAR 2.7.8a
bcftools 1.8 
pysam 0.15.3
shapeit 4.2
WASP 0.3.4
```

**STEP 0: Download cellranger references** If you don't already have the reference files that were used to align your datasets then download them from the 10X Genomics website. For example, if you are analyzing human single cell gene expression and ATAC data you may have used the following references. This tutorial uses GRCh38 cellranger references and GATK bundle resources. If you're using hg19 then make sure to download the files that match your reference.
```
# RNA GRCh38 / hg38
GRCh38-2020-A.premrna

# ATAC GRCh38 / hg38
wget https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-atac-GRCh38-1.2.0.tar.gz

# ATAC GRCh37 / hg19
wget https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-atac-hg19-1.2.0.tar.gz

```

**STEP 0: Download GATK resource bundle files** Download GATK resource bundle files from the [GATK google cloud bucket](https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false)

The following files are required for GATK HaplotypeCaller using GRCh38:
```
resources_broad_hg38_v0_Homo_sapiens_assembly38.dbsnp138.vcf.gz
resources_broad_hg38_v0_Homo_sapiens_assembly38.known_indels.vcf.gz
resources_broad_hg38_v0_1000G_phase1.snps.high_confidence.hg38.vcf.gz
resources_broad_hg38_v0_Mills_and_1000G_gold_standard.indels.hg38.vcf.gz
```
The following additional files are required if you are analyzing single cell ATAC datasets (not needed for the tutorial):
```
resources_broad_hg38_v0_wgs_calling_regions.hg38
resources_broad_hg38_v0_hapmap_3.3.hg38.vcf.gz
```
**STEP 0: Pull the docker containers** Pull the containers from dockerhub. 
```
docker pull p4rkerw/salsa:count_1.0
docker pull broadinstitute/gatk:4.2.0.0
```
**STEP 1: Genotype a single cell gene expression dataset** The library_id variable refers to the name of the output folder generated by cellranger count (or cellranger-atac count) located in the rna_counts (or atac_counts) directory. If you're looking for a smaller dataset to work with in this tutorial, there are lots of options on the [10X Genomics website](https://support.10xgenomics.com/single-cell-gene-expression/datasets). For the purposes of the tutorial, you may consider downloading the coordinate-sorted bam and index for a dataset composed of 1k PBMCs from a healthy donor:
```
# navigate to your project directory
project=$PWD/salsa
wget -P $project/cellranger_rna_counts https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam
wget -P $project/cellranger_rna_counts https://cf.10xgenomics.com/samples/cell-exp/4.0.0/SC3_v3_NextGem_DI_PBMC_CSP_1K/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam.bai
```
**Clone the SALSA github repository**
```
git -C $project clone https://github.com/p4rkerw/SALSA
```
**Usage:**
```
Usage: step1_gatk_genotype.sh [-indomlt]
   -i  | --bam                STR   path/to/input.bam eg. [rna_counts/sample_1/outs/possorted*.bam]
   -n  | --library_id         STR   library_id: eg. [sample_1]
   -d  | --outputdir          STR   output directory name eg. [vcfdir/rna_genotype]
   -o  | --outputvcf          STR   name of output vcf eg. [sample_1.rna.vcf.gz]
   -m  | --modality           STR   sequencing modality for short variant discovery: [rna] [atac]
   -l  | --interval           STR   optional: genotype a single chromosome eg. [chr10]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch GATK container** : Mount the required volumes in an interactive session. The $SCRATCH1 variable designates a temporary file directory. The following command mounts directories with counts generated by cellranger mounted to rna_counts. Users may prefer to mount all the directories required for future steps (see below) when they run the container, but for the sake of simplicity only the required volumes are specified at each step.
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/cellranger_rna_counts:$HOME/rna_counts \
-v $reference/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $reference/gatk:$HOME/gatk_bundle \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it broadinstitute/gatk:4.2.0.0
```
**Genotype an RNA sample**
```
bash SALSA/step1_gatk_genotype.sh \
--bam ${modality}_counts/SC3_v3_NextGem_DI_PBMC_CSP_1K_possorted_genome_bam.bam \
--library_id pbmc \
--outputdir vcfdir/rna_genotype \
--outputvcf pbmc.rna.chr10.vcf.gz \
--interval chr10 \
--modality rna \
--threads 4
```
**(Not required for tutorial) STEP 2: Merge genotypes from the same patient** If you genotyped a paired single cell gene expression and ATAC dataset from a split sample (or a single cell Multiome) you can merge these genotypes into a single vcf. If you're following the tutorial, you can skip this step.
**Usage**
```
Usage: step2_merge_geno.sh [-nabdit]
   -n  | --library_id         STR   library_id: eg. [Control_1]
   -a  | --vcfone             STR   path/to/first.vcf.gz eg. [vcfdir/atac_genotype/Control_1.atac.vcf.gz]. Prioritize annotations from first vcf if there is overlap.
   -b  | --vcftwo                               STR   path/to/second.vcf.gz eg. [vcfdir/rna_genotype/Control_1.rna.vcf.gz]
         -d  | --outputdir                      STR   output directory [vcfdir/joint_genotype]
   -o  | --outputvcf          STR   name of output vcf eg. [Control_1.pass.joint.vcf.gz]
   -i  | --includefiltered          optional: include filtered variants in output vcf. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
**Launch SALSA container**
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Merge two genotypes**
```
bash SALSA/step2_merge_geno.sh \
--library_id sample_1 \
--vcfone vcfdir/$atac_genotype/sample_1.atac.chr10.vcf.gz \
--vcftwo vcfdir/${modalitytwo}_genotype/sample_1.rna.chr10.vcf.gz \
--outputdir vcfdir/joint_genotype \
--outputvcf sample_1.pass.joint.chr10.vcf.gz \
--threads 4
```
**(Recommended) STEP 3: Phase genotype** If you want to perform your analysis with phased genotypes you will need a phased reference. This is not stricly required, but it increases the performance of the WASP variant-realignment and ASEP analysis steps. Download the 1000G phased reference files for either SNV or SNV and indels from ftp.1000genomes.ebi.ac.uk . Navigate to the corresponding directory depending on your selected reference. If you are only analyzing RNA data then select the SNV reference. For ATAC data select the SNV and INDEL reference:

a) SNV only: /vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV </br>
b) SNV and INDEL: /vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL

You will eventually need to download the vcf for every chromosome, but for the purposes of the tutorial just download the SNV reference for chromosome10 to reference/phasing:
```
ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
```
**Usage**
```
Usage: step3_phase_vcf.sh [-nvdolpsitrh]
   -n  | --library_id         STR   library_id: eg. [Control_1]
   -v  | --inputvcf           STR   path/to/input.vcf.gz eg. [vcfdir/joint_genotype/Control_1.pass.joint.vcf.gz]
   -d  | --outputdir          STR   output directory name eg. [vcfdir/phasing]
   -o  | --outputvcf          STR   name of output vcf eg. [Control_1.pass.joint.phase.vcf.gz]
   -l  | --interval           STR   optional: phase a single chromosome eg. [chr22]
   -p  | --hcphase                  optional: recover haplotypecaller physical phasing variants that are not in shapeit reference. Default=[false]
   -s  | --snvonly            STR   use the biallelic_SNV reference for phasing
   -i  | --snvindel           STR   use the biallelic_SNV_and_INDEL reference for phasing
   -r  | --reproduce                optional: run shapeit with a single thread for reproducibility. Default=[false]
   -t  | --threads            INT   number of threads. Default=[1]
   -h  | --help                     show usage
```
The 1000G do not have the same contig style as the cellranger references. Update the 1000G contig style using bcftools in the stage 1 container. For the tutorial we will only do chromosome 10. Mount the directory with the 1000G reference and phase the genotypes. For the tutorial we do not set the --reproduce flag, which is required to set a seed for reproducible results with shapeit4.2. This step is recommended in the full workflow.
**Launch SALSA container**
```
SCRATCH1=/g/scratch
project=/g/salsa
reference=/g/reference
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v $project/vcf_output:$HOME/vcfdir \
-v $project/SALSA:$HOME/SALSA \
-v $reference/phasing/biallelic_SNV:$HOME/phasing \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="/g/scratch" \
--rm -it p4rkerw/salsa:count_1.0
```
**Rename the reference contigs**
```
# rename the contigs in the 1000G reference from 10 to chr10
for i in {1..22} X;do echo "${i} chr${i}";done > /tmp/rename_chrm.txt
inputvcf=ALL.chr10.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
bcftools annotate phasing/$inputvcf --threads 4 --rename-chrs /tmp/rename_chrm.txt -Oz -o $SCRATCH1/$inputvcf
mv $SCRATCH1/$inputvcf phasing/
bcftools index --threads 4 phasing/$inputvcf
```
**Phase an interval**
```
bash SALSA/step3_phase_vcf.sh \
--library_id pbmc \
--inputvcf vcfdir/rna_genotype/pbmc.rna.chr10.vcf.gz \
--outputdir vcfdir/phasing \
--outputvcf pbmc.pass.joint.chr10hcphase.vcf.gz \
--interval chr10 \
--hcphase \
--threads 4 \
--snvonly
```
**(Optional) STEP 4: Annotate vcf with GATK Funcotator** If you would like to annotate your vcf with gnomAD MAF you will first need to download the GATK Funcotator resource following the directions located here: https://gatk.broadinstitute.org/hc/en-us/articles/360035889931-Funcotator-Information-and-Tutorial . The gnomAD resources need to be enabled after download (see GATK instructions on their website). When the resources have been downloaded move the dataSources folder into to the reference directory (eg. [reference/funcotator_dataSources.v1.6.20190124])

Mount the reference directory to the docker container and annotate the vcf

```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v path/to/refdata-cellranger-atac-GRCh38-1.2.0:$HOME/atac_ref \
-v path/to/vcf_output:$HOME/vcfdir \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/reference:$HOME/reference \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
interval=chr10
bash diabeticKidney/allele_specific_analysis/step4_gatk_anno_vcf.sh \
--library_id $library_id \
--inputvcf vcfdir/phasing/$library_id.pass.joint.${interval}hcphase.vcf.gz \
--outputdir vcfdir/funcotation \
--outputvcf $library_id.pass.joint.${interval}hcphase.funco.vcf.gz \
--output_table $library_id.pass.joint.${interval}hcphase.formatted.csv \
--modality atac \
--threads 4 \
--funcotation reference/funcotator_dataSources.v1.6.20190124g
```

**(Recommended) STEP 5:** Use barcode celltype annotations to filter the coordinate-sorted cellranger bam using the CB tag. This step is required if you want to analyze cell-specific or single cell allelic effects. The barcode annotation file has three columns where the first column is the barcode, the second column is the library_id, and the third column is the celltype annotation. If you have analyzed your dataset using Seurat you could generate a barcode annotation csv in R as follows (the Seurat package is not included in the SALSA containers):
```
library(Seurat)
library(stringr)
celltype <- Idents(seurat_obj)
barcode <- str_split(rownames(seurat_obj@meta.data), pattern="-", simplify=TRUE)[,1]
library_id <- seurat_obj@meta.data$orig.ident
anno <- data.frame(barcode=barcode, library_id=library_id, celltype=celltype)
write.csv(anno, file="rna_barcodes.csv", row.names=FALSE, quote=FALSE)
```
Mount the barcodes directory and filter the cellranger bam using the 10X Genomics subset-bam utility. For the purposes of the tutorial, we will only filter chromosome 10. The --validate flag runs GATK ValidateSamFile on the output to ensure the bam file meets specifications. To filter an ATAC coordinate-sorted bam change the modality and library_id variables as needed.
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/cellranger_rna_counts:$HOME/rna_counts \
-v path/to/cellranger_atac_counts:$HOME/atac_counts \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/barcodes:$HOME/barcodes \
-v path/to/project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
interval=chr10
modality=rna
bash diabeticKidney/allele_specific_analysis/step5_filterbam.sh \
--library_id $library_id \
--validate \
--inputbam ${modality}_counts/$library_id/outs/possorted_*.bam \
--modality $modality \
--interval $interval \
--barcodes barcodes/${modality}_barcodes.csv \
--threads 4 \
--outputdir project/wasp_${modality} \
--outputbam $library_id.bcfilter.${interval}.bam 
```

**STEP 6: Perform variant-aware realignment with WASP** This step takes a genotyped vcf and performs variant-aware realignment on a coordinate-sorted and indexed bam file with WASP. WASP is a tool to perform unbiased allele-specific read mapping and you can read more about it here: https://github.com/bmvdgeijn/WASP . For the purposes of the tutorial, we will only analyze chromosome 10. For RNA analysis, this step requires a STAR index of the cellranger reference. A STAR index can be built ahead of time using the following command:
```
STAR
--runMode genomeGenerate \
--runThreadN $threads \
--genomeDir rna_ref/star \
--genomeFastaFiles rna_ref/fasta/genome.fa \
--sjdbGTFfile rna_ref/genes/genes.gtf
```

Alternatively, these references will be built at runtime and placed in the $SCRATCH directory. To perform variant aware realignment on a single cell gene expression dataset with WASP run:
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v path/to/vcf_output:$HOME/vcfdir \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
modality=rna
interval=chr10
bash diabeticKidney/allele_specific_analysis/step6_wasp.sh \
--inputvcf vcfdir/funcotation/$library_id.pass.joint.hcphase.funco.vcf.gz \
--inputbam project/wasp_${modality}/$library_id.bcfilter.bam \
--outputdir project/wasp_${modality}/joint_genotype \
--outputbam $library_id.phase.${interval}wasp.bam \
--genotype joint \
--stargenome rna_ref/star \
--library_id $library_id \
--modality $modality \
--threads 4 \
--isphased \
--interval $interval
```

Similarly, if you are analyzing ATAC data you will need to create an index for bwa. A bwa index can be built ahead of time using the following command:
```
bwa index atac_ref/fasta/genome.fa 
```
To perform variant aware realignment on a single cell ATAC dataset with WASP run:
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/refdata-cellranger-atac-GRCh38-1.2.0:$HOME/atac_ref \
-v path/to/vcf_output:$HOME/vcfdir \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
modality=atac
interval=chr10
bash diabeticKidney/allele_specific_analysis/step6_wasp.sh \
--inputvcf vcfdir/funcotation/$sample.pass.joint.${interval}hcphase.funco.vcf.gz \
--inputbam project/wasp_${modality}/$library_id.bcfilter.${interval}.bam \
--outputdir project/wasp_${modality} \
--outputbam $library_id.phase.${interval}wasp.bam \
--genotype joint \
--library_id $library_id \
--modality $modality \
--threads 4 \
--isphased
```

**STEP 7: Obtain allele-specific read counts with GATK** This step will filter a phased and genotyped vcf for heterozygous SNV to perform allele-specific counting in a coordinate-sorted and indexed bam file after WASP realignment. There are multiple options for count table outputs. The --pseudobulk option will group all barcodes together to perform allele-specific counting. This is analogous to bulk RNA-seq. The --celltype_pseudobulk option will use the barcode annotations to split the bam into cell-type-specific bam files before performing allele-specific counting. The --single_cell_counts option will split the bam file into individual single cell bam files and perform allele-specific counting. If your input vcf is phased you can select the --isphased option to add a phased genotype to the count tables.

To perform pseudobulk, celltype pseudobulk, and single cell allele-specific counts with a phased input vcf:
```
SCRATCH1=path/to/scratch
docker run \
--workdir $HOME \
-v $HOME:$HOME \
-v path/to/GRCh38-2020-A.premrna:$HOME/rna_ref \
-v path/to/refdata-cellranger-atac-GRCh38-1.2.0:$HOME/atac_ref \
-v path/to/vcf_output:$HOME/vcfdir \
-v path/to/SALSA:$HOME/SALSA \
-v path/to/barcodes:$HOME/barcodes \
-v path/to/project:$HOME/project \
-v $SCRATCH1:$SCRATCH1 \
-e SCRATCH1="path/to/scratch" \
--rm -it p4rkerw/salsa:count_1.0

library_id=sample_1
modality=rna
interval=chr10
bash diabeticKidney/allele_specific_analysis/step7_gatk_alleleCount.sh \
--inputvcf vcfdir/funcotation/$library_id.pass.joint.${interval}hcphase.funco.vcf.gz \
--inputbam project/wasp_${modality}/$library_id.phase.${interval}wasp.bam \
--outputdir project/wasp_${modality}/counts \
--barcodes barcodes/${modality}_barcodes.csv \
--genotype joint_genotype \
--library_id $library_id \
--modality $modality \
--pseudobulk_counts \
--single_cell_counts \
--celltype_counts \
--threads 4 \
--interval $interval \
--isphased
```








